ml4h/models/model_factory.py

```diff
@@ -0,0 +1,282 @@
+    :param custom_dict: Dictionary of custom objects needed to load a serialized model
+    :param u_connect: dictionary of input TensorMap -> output TensorMaps to u connect to
+    """
+    encoder_block_functions = {tm: identity for tm in tensor_maps_in}  # Dict[TensorMap, Block]
+    for tm in tensor_maps_in:
+        for encode_block in encoder_blocks:  # TODO: just check that it is a block?,
+            if isinstance(encode_block, Block):
+                encoder_block_functions[tm] = compose(encoder_block_functions[tm], encode_block(tensor_map=tm, **kwargs))
+            elif encode_block in BLOCK_CLASSES:
+                encoder_block_functions[tm] = compose(encoder_block_functions[tm], BLOCK_CLASSES[encode_block](tensor_map=tm, **kwargs))
+            elif encode_block.endswith(f'encoder_{tm.name}.h5'):  # TODO: load protobuf models too
+                serialized_encoder = load_model(encode_block, custom_objects=custom_dict, compile=False)
+                encoder_block_functions[tm] = compose(encoder_block_functions[tm], ModelAsBlock(tensor_map=tm, model=serialized_encoder))
+                break  # Don't also reconstruct from scratch if model is serialized, hd5 models must precede BLOCK_CLASS keys
+
+    merge = identity
+    for merge_block in merge_blocks:
+        if isinstance(merge_block, Block):
+            merge = compose(merge, merge_block(**kwargs))
+        else:
+            merge = compose(merge, BLOCK_CLASSES[merge_block](**kwargs))
+
+    decoder_block_functions = {tm: identity for tm in tensor_maps_out}
+    for tm in tensor_maps_out:
+        for decode_block in decoder_blocks:
+            if isinstance(decode_block, Block):
+                decoder_block_functions[tm] = compose(decoder_block_functions[tm], decode_block(
+                    tensor_map=tm,
+                    u_connect_parents=[tm_in for tm_in in tensor_maps_in if tm in u_connect[tm_in]],
+                    parents=tm.parents,
+                    **kwargs,
+                ))
+            elif decode_block in BLOCK_CLASSES:
+                decoder_block_functions[tm] = compose(decoder_block_functions[tm], BLOCK_CLASSES[decode_block](
+                    tensor_map=tm,
+                    u_connect_parents=[tm_in for tm_in in tensor_maps_in if tm in u_connect[tm_in]],
+                    parents=tm.parents,
+                    **kwargs,
+                ))
+            elif decode_block.endswith(f'decoder_{tm.name}.h5'):
+                serialized_decoder = load_model(decode_block, custom_objects=custom_dict, compile=False)
+                decoder_block_functions[tm] = compose(decoder_block_functions[tm], ModelAsBlock(tensor_map=tm, model=serialized_decoder))
+                break
+
+    return _make_multimodal_multitask_model_block(encoder_block_functions, merge, decoder_block_functions, u_connect)
+
+
+def _make_multimodal_multitask_model_block(
+        encoder_block_functions: Dict[TensorMap, Block],
```

Comment:
If we want one block per tensormap, can we have a `BlockFromBlocks` class that chains blocks together?
That way you could have an encoder be a series of 3 residual blocks

Reply:
Yes we could but it also works currently with lists of Blocks via the 
```
def compose(f, g):
    return lambda x, y: g(f(x, y), y)
```
then during model building calling: `compose(encoder_block_functions[tm], encode_block(tensor_map=tm, **kwargs))`